pypit/processimages.py

```diff
@@ -76,7 +76,7 @@ class ProcessImages(object):
     oscansec : list
     """
     def __init__(self, file_list, spectrograph=None, settings=None, det=1, user_settings=None,
-                 datasec_img=None, bpm=None):
+                 datasec_img=None, bpm=None, spectro_class=None):
```

Comment:
I took a look through the code, and it appears that we no longer need to pass `spectrograph` as an argument (right?). So, I think we just pass `spectro_class` instead of spectrograph, and make it a requirement (i.e. not an optional argument).

Also, we then wouldn't need the import on line 17: `from pypit.spectrographs import spectro_utils`

Furthermore, you could then remove lines 114-121

Reply:
@jhennawi will insist that the classes *not* require this spectro_class as an
argument and instead that it can be derived from the input file.  I am bending to
his will for now..

Reply:
I (respectfully!) disagree with this, since the necessary argument is just one extra line of code (which, while being a class, only requires a text argument - the spectrograph name - as input), and will avoid having to re-instantiate the class (which will eventually contain the spectrograph settings and other functions as well). In my opinion, we're not using the advantage of a class by doing this, and I can't see any positive to regenerate the `spectro_class`.

```
spectro_class = spectro_utils.load_spec_class(spectrograph=spectrograph)
ProcessImages(filelist, spectro_class, **kwargs)
```

Reply:
@jhennawi --  Please enunciate why this is such a great burden to you.

It is making me jump through extra hoops to avoid having it
be a required argument.

Reply:
Again, I'm not following why this extra installation of the spectro_class is necessary, unless you can convince me that it results in dramatically more elegant, modular, or easy to use code.  

Here is why I find passing objects like this into codes insidious. 

Consider the following code example from PYPIT: 

ProcessImages(file_list, det = 1) (or whatever child of process images, e.g. ArcImage)  

----
Inside Process images: 
----
hdr = hdufits(file_list[0])
spectro_class = get_instrument(hdr)

------
Inside get_instrument (to Ryan's dread!, but only appearing here)
------
instrument = hdr['INSTRUME']

if instrument == 'keck_lris_blue': 

elif instrument == ....

if instrument is not recognized, a generic spectro_class is returned with some sensible default assumptions. 

---- 
Process Images does whatever it needs using the instrument specific spectro class. In this case, that is simply read in the image correctly dealing with instrument specific idiosyncrasies. 
----
Process images returns 
return msarc, hdr 

wavetilts = WaveTilts(msarc, hdr, det = 1, parset = None) 


(I'd actually prefer if WaveTilts took the file_list rather than an image so that we would not have to pass the header in here, and that the extra call to the ArcImage class could be avoided.  But this is an argument I appear to be losing with @profxj )

Note that here waveTilts determines the default parset in a keck_lris_blue 1200 line grating specific way because it uses information in the hdr. parset is an optional argument we use only if we want to alter the defaults. 

If I want to run WaveTilts with a different set of parameters then the defaults for that function and instrument keck_lris_blue I would the modified sequence

wavetilts = WaveTilts(None, hdr)
changed_parset = wavetilts.default_parset

# make my changes to the parset
changed_parset['threshold']=XXX

# Now run again via instantiaion
wavetilts_new = WaveTilts(msarc, hdr, parset = wavetilts_parset)

# Or more simply type, using the original instantiation above
wavetilts.run(parset = changed_parset)

The wavetilts object would provide me with parset it used via the attribute
parset_used = wavetilts.parset_used
 
# Now later on down the line, I want to make the 2-d wavelength image, which builds on the tilts output.  For that I would run: 

waveimage  = WaveImage(tslits_dict['slitpix'], wavetilts.tilts, wavecalib.wv_calib)

Essentially, for any class that does something in an instrument specific way, either the filename or the hdr needs to be passed in.  You may think it is cumbersome to pass a hdr around. But notice that the user never has to instantiate any extra unnecessary object. It happens behind the scenes. I also never need to pass this spectro_class object to any other class during the control flow. 

The reason why I think passing in a hdr is superior to passing in an object is that the user doesn't know what the spectro_class object is. They don't know how to access or modify its attributes and whether that is required.  They don't know whether upstream routines modified its attributes or contents in way that is required for downstream operations  to run correctly. They don't know whether you are grabbing things out of that spectro_class,  e.g. if it were required to pass the spectro_class into say WaveImage that could mean WaveImage depends on hidden input parameters that are not explicitly declared. If the user decides they want to skip a step in the reduction, they might be confused about whether they need to load something into that spectro_class object to get the next step to work. 

All of these questions and doubts that you introduce by passing complex mutable things around represent issues that myself and my team members have encountered while trying to develop and use pypit. It may be clear to the developers that the spectro_class is an immutable container for instrument specific things, but this is much less clear to the user. 

Now you can argue that the calibration etc. classes can be doing instrument specific things "behind the scenes" as well based on the header information. That is fine and we can state that explicitly in the docs. "PYPIT algorithms have instrument specific defaults that will be automatically used for your instrument based on the header information. If you want alter them you  can explicitly modify the default_parset and input it into the relevant code." 

The main point is that a header and a filename are very familiar and immutable things (technically immutable not formally) that everyone understands. There are no hidden inputs and the coding model is completely transparent.  

The only obvious cons I see are that: 

1. Sometimes headers don't have enough information (but we will know what instrument these are and this problem has to be dealt with irrespective of the coding model). 

2. Sometimes headers get corrupted, but that is easy enough to fix by grabbing a non-corrupted header from the same setup etc. 

I should say that I'm not that fixed on the hdr idea. A dictionary would also work as long as we agreed that codes are not allowed to alter its contest (unfortunately immutable dictionaries in pypit are complicated to create). 

I'm eager to hear your thoughts on this, and am willing to concede to your alternative model if you can provide an explicit code example where this header model fails. @profxj you complained about jumping through hoops. Can you provide a specific code example?

Joe

Reply:
I think the cleanest example of where this type of coding is much cleaner is in `ararc.setup_param`. Previously, there was a long
```
if instrumentName:
    foo
elif otherInstrumentName:
    bar
elif ...
```
case there (and there are many others throughout the code too). Instead, we now have a single call to the `spectro_class`, which gets the information from a class in a single file (e.g. keck_deimos.py).

This means that if you (or anyone else) wanted to add a new instrument, you don't need to search through the entire code to find (and catch) all of those `if/elif instrumentName` case statements. You would *only* need to open up a file called `telescopeName_instrumentName.py` and make changes to that file *only*. (You would also need to add two lines of code to `spectro_utils.py`). So, from a developer point of view (and for describing how to do this in the docs) this new approach is much cleaner and clearer.

A further advantage of this approach is that all children will inherit the defaults from their parent. So, if we wanted to change the default way of doing some operation (that is specific to the instrument) we would only need to change it in one file (i.e. the parent class in `spectroclass.py`) and all of the children will inherit that change.

Furthermore, this new class acts as a container with all of the settings of this particular instrument (e.g. the dispersion axis), how to read in the data, etc.. So, *everything* that you need to develop a new instrument is in just one file.

The only thing that is needed to create one of these classes is a call to spectroclass with a string that has the name of the spectrograph.

In fact, the idea behind this approach was largely developed because of one of the telecons where you expressed how difficult it was to know how to change and/or add a setting to PypIt. The approach proposed here would mean that you add two lines of code to `spectro_utils.py`, and everything else is manually defined and coded by yourself in a file called `telescopeName_instrumentName.py`.

All that said, I'm very much open to a better approach (I just don't see it!).